src/Thruway/Role/Subscriber.php

```diff
@@ -219,13 +251,49 @@ public function subscribe(ClientSession $session, $topicName, callable $callback
             "options"    => $options,
             "deferred"   => $deferred
         ];
-
+        
         array_push($this->subscriptions, $subscription);
-
+        
         $subscribeMsg = new SubscribeMessage($requestId, $options, $topicName);
         $session->sendMessage($subscribeMsg);
 
         return $deferred->promise();
     }
-
+	
+    /**
+     * process unsubscribe
+     * @param ClientSession $session
+     * @param type $topicName
+     * @return boolean
+     */
+    public function unsubscribe(ClientSession $session, $topicName)
+    {
+        $requestId = Utils::getUniqueId();
+        $subscriptionId = false;
+        $deferred  = new Deferred();
+        
+        foreach ($this->subscriptions as $i => $subscription) {
+            if ($subscription["topic_name"] == $topicName) {
+                $subscriptionId = $subscription["subscription_id"];
+                
+                $this->subscriptions[$i]["unsubscribed_request_id"] = $requestId;
+                $this->subscriptions[$i]["unsubscribed_deferred"] = $deferred;
+            }
+        }
```

Comment:
If we take a look at Autobahn|JS client, the `subscription` method returns an object representing the topic. When unsubscribing, this object must be given to the `unsubscribe` method (the topic name is not given.) Maybe it is done this way to handle the situation you just described (the object Autobahn creates must contain the subscription id or something.) Maybe we should do something similar?

The `ClientSession::subscribe` method could return a custom object (let's call it `Subscription`) that wraps a `Deferred` object so that the `then` method could still be available. When the router responds, the `Subscription` object is updated (if the subscription succeeded, the attribute `subscription_id` is filled; if failed, nothing special happens; in both cases, the promise is either resolved or rejected.) The `ClientSession::unsubscribe` method would then be rewritten to accept a `Subscription` object in argument (and of course the `Subscriber::unsubscribe` would be rewritten too.)

What do you think?